Relocate commands into discrete classes

tim-savage opened the pull request:
This pull request moves all commands into discrete classes for each command, in a new package named 'commands'. A new top level class, 'CommandManager', has been created which instantiates each command, and provides some common convenience methods. A field has been declared in the plugin main class that stores a reference to the instance of the CommandManager, which itself is instantiated and assigned to this field at the very end of the onEnable method. Each command is self registering with the server through the setExecutor method invoked in their respective constructors.

A small number of classes, methods and fields were required to have their visibility elevated to public, in order to allow access to the new classes due to the fact they reside in a separate package. These changes in visibility have been kept to the absolute minimum necessary, and no other changes have been made. 

The intention of this change was to remove a large amount of code from the plugin main class, and breaking the commands into separate classes should enable easier debugging and enhancements, as the relevant code relating to a specific command now resides in its own class. Nearly 2k lines were extracted from the plugin main class, and most new classes contain their onCommand logic within a block which is viewable without scrolling. This code remains virtually unchanged from the original, except where changes were necessary, although I did address potential NPE's in a couple of places. All other code that produces warnings was left in place, to be addressed (and tested) at a later date.

As an added bonus, I have taken the liberty to add tab completion where appropirate to the commands that take additional arguments. Player commands that take another player name as an argument are limited to currently online players to avoid disclosing the entire list of known players in the lifetime of the server to any player. Some admin commands dispense with this limitation for the convenience of those with administrator privileges.

I also would like to add that this change could faciliate the migration to using subcommands in the future, if so desired.

Jikoo commented on src/main/java/me/ryanhamshire/GriefPrevention/CommandManager.java:
```diff
@@ -0,0 +1,218 @@
+        for (Player otherPlayer : plugin.getServer().getOnlinePlayers())
+        {
+            // if other player name matches prefix, add name to results list
+            if (otherPlayer.getName().startsWith(prefix))
```
Should probably use `org.bukkit.util.StringUtil#startsWithIgnoreCase(String, String)`.

Jikoo replied:
Should also be verifying `Player#canSee(Player)` from sender if sender is a player to not give away online status for vanished players.

Jikoo commented on src/main/java/me/ryanhamshire/GriefPrevention/CommandManager.java:
```diff
@@ -0,0 +1,218 @@
+    public List<String> matchOfflinePlayers(final CommandSender sender, final String prefix, final boolean excludeSelf)
+    {
+        List<String> results = new LinkedList<>();
+        for (OfflinePlayer otherPlayer : plugin.getServer().getOfflinePlayers())
```
This will cause a prohibitive amount of IO on servers with larger numbers of players. Combined with GP targeting Spigot (i.e. no async tab completion, etc.) it's a recipe for disaster.

Jikoo commented on src/main/java/me/ryanhamshire/GriefPrevention/PlayerData.java:
```diff
@@ -52,7 +52,7 @@
 
     //radius for restore nature fill mode
-    int fillRadius = 0;
+    public int fillRadius = 0;
```
Personally would like to see freshly exposed values use setter/getter. This goes for all the values in PlayerData being exposed, but it's personal preference and Robo may feel differently.

Jikoo commented on src/main/java/me/ryanhamshire/GriefPrevention/commands/AbandonTopLevelClaimCommand.java:
```diff
@@ -0,0 +1,38 @@
+
+import java.util.Objects;
+
+public final class AbandonTopLevelClaimCommand implements CommandExecutor
```
Why does this class exist when it could just be a command name (not label) check in the AbandonClaimCommand? Loads of duplicate code with a single boolean value difference that could just be `command.getName().equals("AbandonTopLevelClaim")` (presumably need caps to match plugin.yml for equality to command name, not sure)

Jikoo commented on src/main/java/me/ryanhamshire/GriefPrevention/commands/AccessTrustCommand.java:
```diff
@@ -0,0 +1,60 @@
+import java.util.List;
+import java.util.Objects;
+
+public final class AccessTrustCommand implements CommandExecutor, TabCompleter
```
Trusts should also be merged under a single command executor.

An aside: `TabExecutor` would reduce your own boilerplate, but it's done now.

Jikoo commented on src/main/java/me/ryanhamshire/GriefPrevention/commands/GPBlockInfoCommand.java:
```diff
@@ -0,0 +1,50 @@
+public final class GPBlockInfoCommand implements CommandExecutor
+{
+    // reference to plugin main class
+    @SuppressWarnings({"FieldCanBeLocal", "unused"})
```
Probably should just make it a local field - suppression is generally bad unless there's no alternative. I know you did it for consistency, but it's unnecessary.

Jikoo commented on src/main/java/me/ryanhamshire/GriefPrevention/commands/PermissionTrustCommand.java:
```diff
@@ -0,0 +1,58 @@
+        //requires exactly one parameter, the other player's name
+        if (args.length != 1) return false;
+
+        plugin.handleTrustCommand(player, null, args[0]);  //null indicates permissiontrust to the helper method
```
When the trust commands are revisited we should probably revisit passing null for no reason. This was some legacy code from before PermissionTrust had a ClaimPermission I believe. If we're exposing it it should probably be improved on. That said, it could just be cut and pasted as a helper method in a unified trust command.

Jikoo reviewed (changes requested):
While I like the idea, this is a prohibitively large diff. I know I missed several cases because I forgot by the time I ended up scrolling to them. IMO this should be done in batches of related commands over multiple PRs but I get why you didn't.

Big blocker for me is offline player tab completion - not a fan. Loading all offline players is flat out dangerous. I used to manage a server where the act of loading all offline players (and we only had something like 6k total players all time, we were not large) took several seconds. Because GP supports CB as minimum required software, tab completion is done on the main thread. This means tab completion for those admin commands will hang the server every time it's done, and modern tab completion fires very fast.

Any helper methods from the `GriefPrevention` class should be helper methods inside a unified respective command where applicable rather than expose them and create loads of duplicate code per command.

tim-savage committed:
rename CommandManager -> CommandHandler

renamed for consistency with existing codebase
commented out all command instantiation statements, to be uncommented as new classes are added.

tim-savage committed:
delete matchOfflinePlayers methods

tim-savage committed:
remove OfflinePlayer import

tim-savage committed:
add method to check configuration variable for tab completion

this placeholder stub will be used by all command tab completer methods to test if tab completion has been enabled/disabled in config.yml

tim-savage committed:
make static fields public for use by classes in command package.

tim-savage committed:
declare abandonClaimHandler public for access from commands.AbandonClaimCommand class

tim-savage committed:
initial commit of AbandonAllClaimsCommand class

tim-savage committed:
initial commit of AbandonClaimCommand class

this class handles both the abandonclaim and abandontoplevelclaim commands

tim-savage committed:
initial commit of AbandonAllClaimCommand class

this class handles both the abandonclaim and abandontoplevelclaim commands

tim-savage committed:
correct javadocs for matchOnlinePlayers methods

tim-savage committed:
correct javadocs for matchOnlinePlayers methods

tim-savage committed:
Merge remote-tracking branch 'origin/relocate_commands' into relocate_commands

tim-savage committed:
migrate trust commands

tim-savage committed:
migrate adjust bonus claim blocks commands

tim-savage committed:
migrate admin/basic claims commands

tim-savage committed:
migrate buy/sell claim blocks commands

tim-savage committed:
migrate separate/unseparate commands

tim-savage committed:
migrate ignore/unignore player commands

tim-savage committed:
migrate restore nature player commands

tim-savage committed:
migrate subclaim commands

tim-savage committed:
migrate delete claim commands

tim-savage committed:
migrate claim command

tim-savage committed:
migrate claimbook command

tim-savage committed:
migrate claims list command

tim-savage committed:
migrate claim explosions command

tim-savage committed:
migrate extendclaim command

tim-savage committed:
migrate give pet command

tim-savage committed:
migrate block info command

tim-savage committed:
migrate reload command

tim-savage committed:
migrate ignore claims command

tim-savage committed:
migrate setaccruedclaimblocks command

tim-savage committed:
migrate siege command

tim-savage committed:
migrate siege command

tim-savage committed:
migrate transfer claim command

tim-savage committed:
migrate trapped command

tim-savage committed:
migrate unlock drops command

tim-savage committed:
remove onCommand from main class, and enable all commands in CommandHandler
